adcm_aio_client/core/requesters.py

```diff
@@ -22,20 +40,129 @@ def as_dict(self: Self) -> dict: ...
+
+        return data
+
+    def as_dict(self: Self) -> dict:
+        if not isinstance(data := self._prepare_json_response(), dict):
+            message = f"Expected a dict, got {type(data)}"
+            raise RequesterResponseError(message)
+
+        return data
 
-    async def post(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+    def _prepare_json_response(self: Self) -> Json:
+        try:
+            data = self.response.json()
+        except JSONDecodeError as e:
+            message = "Response can't be parsed to json"
+            raise RequesterResponseError(message) from e
 
-    async def patch(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+        if "results" in data:
+            data = data["results"]
 
-    async def delete(self: Self, path: str) -> AsyncGenerator[RequesterResponse]: ...
+        return data
 
 
-class Session: ...
+def _get_error_class(status_code: int) -> type[ResponseError]:
+    first_digit = status_code // 100
+    match first_digit:
+        case 5:
+            return ServerError
+        case 4:
+            match status_code:
+                case 400:
+                    return BadRequestError
+                case 401:
+                    return UnauthorizedError
+                case 404:
+                    return NotFoundError
+                case 409:
+                    return ConflictError
+
+    return ResponseError
+
+
+def convert_exceptions(func: Callable) -> Callable:
+    async def wrapper(*arg: Iterable, **kwargs: dict) -> httpx.Response:
+        response = await func(*arg, **kwargs)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            raise _get_error_class(status_code=response.status_code) from e
+
+        return response
+
+    return wrapper
 
 
 class DefaultRequester(Requester):
-    def __init__(self: Self) -> None: ...
+    __slots__ = ("api_root", "session")
+
+    def __init__(self: Self, url: str, credentials: Credentials | None = None) -> None:
+        self.api_root = urljoin(url, "/api/v2/")
+        self.client = httpx.AsyncClient()
+        if credentials:
+            get_event_loop().run_until_complete(self.login_as(credentials=credentials))
+
+    async def login_as(self: Self, credentials: Credentials) -> None:
+        login_url = urljoin(self.api_root, "login/")
+        response = await self._do_request(self.client.post(url=login_url, data=credentials.dict()))
+
+        if response.status_code != 200:
+            message = f"Authentication error: {response.status_code} for url: {login_url}"
+            raise SessionError(message)
+
+    async def get(self: Self, path: str, query_params: dict | None = None) -> HTTPXRequesterResponse:
+        return await self.request(method="get", path=path, params=query_params or {})
+
+    async def post(self: Self, path: str, data: dict, **kwargs: dict) -> HTTPXRequesterResponse:
```

Comment:
`kwargs` is not a `dict` or not only a dict

lets remove kwargs for a time being? those are implementation specific

Reply:
removed